src/libopensc/cwa-dnie.c

```diff
@@ -905,9 +814,20 @@ static int dnie_wrap_apdu(sc_card_t * card, sc_apdu_t * apdu)
 int dnie_transmit_apdu(sc_card_t * card, sc_apdu_t * apdu)
 {
 	int res = SC_SUCCESS;
-	res = dnie_wrap_apdu(card, apdu);
-	if (res <= 0) return res;
-	return sc_transmit_apdu(card, apdu);
+	sc_context_t *ctx;
+	cwa_provider_t *provider = NULL;
+	ctx=card->ctx;
+	provider = GET_DNIE_PRIV_DATA(card)->cwa_provider;
+	if ((provider->status.session.state == CWA_SM_ACTIVE) &&
+		(card->sm_ctx.sm_mode == SM_MODE_TRANSMIT)) {
+		res = dnie_wrap_apdu(card, apdu);
+		LOG_TEST_RET(ctx, res, "Error in dnie_wrap_apdu process");
+		res = cwa_decode_response(card, provider, apdu);
+		LOG_TEST_RET(ctx, res, "Error in cwa_decode_response process");
```

Comment:
This is again a bad choice of names. dnie_wrap_apdu does not only wrapping, it encodes and transmits the APDU. The response is then decoded after transmission.